### ISSUE TITLE
Add a hook function to convert a MIR into a NIR instead of relying on internal logic

### ISSUE THREAD

**Issue by denevers** [NRCan/GSIP#63]
Currently, when a MIR (/info/) is requested from GSIP, it can "know" what NIR we are referring to by replacing /info/ with /id/.  This assumption is embedded in GSIP architecture.  This prevents GSIP MIR template to be used used with arbitrary PID.  A use case proposed by @dblodgett-usgs  is to have the NIR managed by a PID that would redirect to GSIP to do the MIR job (pull data from the storage and build the page). Right now it's not possible because GSIP will always assume that the MIR is a representation of a NIR built by replacing /info/ with /id/.

This logic is embedded in Information.java

```java
public  String getIdResource(UriInfo uriInfo)
	{
		// get the /id/ resource and convert to persistant
		String persistentUri = Manager.getInstance().getConfiguration().getParameterAsString(PERSISTENT_URI, BASE_URI);
		// this one is called by /info/, but the real NIR is /id/
		StringBuilder infoUri = new StringBuilder(persistentUri + "/id");
		boolean first = true;
		for(PathSegment segment: uriInfo.getPathSegments())
		{
			if (first) {first = false; continue;}
			infoUri.append("/"+segment.getPath());
		}
		//Logger.getAnonymousLogger().log(Level.INFO, "processed URI:" + uriInfo.getPath());
		//Logger.getAnonymousLogger().log(Level.INFO,"/id/ URI:" + infoUri);
		return infoUri.toString();
		// 
	}
```
(https://github.com/NRCan/GSIP/blob/master/src/nrcan/lms/gsc/gsip/Information.java)

One solution would be to add a hook to a scripting language (Nashorn) in python or javascript  - or a regexp ? that user could configure to provide custom MIR->NIR  transformation.  Absence of hook would result in current transformation logic (info -> id)

**Comment by denevers** [comment]
We could also just use freemarker...

**Comment by denevers** [comment]
commit 03d070e7922fe10b673928a1d969230cbc20c32e

used a freemarker template.

1- GSIP will check if a template named `_mir2nir.ftl' exists.  If there is, it will pass the MIR URL and expect a string back. The string is considered the NIR to be used to query the RDF repo.  If the template is not present, it's GSIP business as usual.
2- In the current commit , I have an example template 

```freemarker
<#-- this template is passed a variable calle mir.  the returning template must
contain a new URI representing the associated NIR.
Note that freemarker comments, like this one, are not serialized -->
${mir?replace("/info/","/id/")}
<#-- using https://freemarker.apache.org/docs/ref_builtins_string.html#ref_builtin_replace
WARNING, demonstration purpose, this is not safe as it assumes /info/ appears only once 
WARNING, it also has no clue about persistent URO in debug mode, etc..-->
```
this is supposed to do what GSIP normally does BUT, there is an important caveat.

the template won't address dev / prod URI because it can't assume where the NIR really is.  One must address this manually.  I could add a replacement for localhost to geoconnex for example (TODO)

The example file I uploaded has extension .ftlx (so, GSIP assume there are no mir2nir template in this commit)

could be improved by example by passing query parameters (so, one could pass `?nir=whatev`) and catch this in the template

**Comment by denevers** [comment]
freemarker string functions:

https://freemarker.apache.org/docs/ref_builtins_string.html

**Comment by dblodgett-usgs** [comment]
It works! http://geoconnex.us/chyld-pilot/id/hu/041504081604
 Still need to forward `f=` and `callback=` from the pid server.

**Comment by denevers** [comment]
closing this (commit 03d070e) and merged into main
